# Praktikum IF3270 2023/2024

Tujuan praktikum IF3270 Pembelajaran Mesin:
1.   Peserta memahami rangkaian proses analitik data menggunakan pendekatan pembelajaran mesin. 
2.   Peserta memahami bahwa proses pengembangan model pembelajaran mesin juga ditentukan dari kualitas data, penanganan data, dan penentuan algoritma serta hyperparameter-nya; tidak cukup hanya dengan memastikan implementasi algoritma berjalan tanpa kesalahan.
3.   Peserta mampu menginterpretasikan hasil dari evaluasi model dalam proses analitik menggunakan pendekatan pembelajaran mesin.

Praktikum dilaksanakan secara berkelompok. Setiap kelompok terdiri atas 2 mahasiswa. Perhatikan bahwa terdapat berkas yang harus dikumpulkan saat keberjalanan praktikum untuk bagian A (25 April 2024, pukul 12.00 WIB) dan berkas yang dikumpulkan setelah waktu praktikum selesai untuk bagian B (25 April 2024, pukul 21.00 WIB).

Disediakan data yang sudah dibagi menjadi data latih (`df_train`), data validasi (`df_val`), dan data uji (`df_test`).

**Bagian 1**: (batas waktu: 25 April 2024, 12.00 WIB)

1. Buatlah _baseline_ dengan menggunakan model _logistic regression_.
2. Lakukan analisis data terkait hal berikut:
    - _duplicate value_,
    - _missing value_,
    - _outlier_,
    - _balance of data_.
3. Jelaskan rencana penanganan yang ada pada poin 2.
4. Jelaskan teknik _encoding_ yang digunakan terhadap data yang disediakan apabila dilakukan, disertai dengan alasan.
5. Buatlah desain eksperimen dengan menentukan hal berikut:
    - tujuan eksperimen,
    - variabel dependen dan independen,
    - strategi eksperimen,
    - skema validasi.
    
**Bagian 2**: (batas waktu: 25 April 2024, 21.00 WIB)

6. Implementasikan strategi eksperimen dan skema validasi yang telah ditentukan pada poin 5.
7. Berdasarkan hasil prediksi yang dihasilkan, buatlah kesimpulan analisis **hasil diabetes**.

---
Catatan:
- Jika terdapat perubahan jawaban pada poin 1—5 (contoh: perbedaan penanganan _outlier_), jelaskan pada laporan mengenai jawaban sebelum, jawaban sesudah, dan alasan pengubahan jawaban.
- Eksperimen dapat berupa penggantian model klasifikasi, pengaturan hyperparameter, model stacking, grid search, oversampling, undersampling, dan lain sebagainya. Semakin variatif eksperimen yang dilakukan, semakin baik.

## Dataset
`diabetes.csv` merupakan dataset yang telah dimodifikasi dari [Diabetes Health Indicators Dataset](https://www.kaggle.com/datasets/alexteboul/diabetes-health-indicators-dataset/) sebagai kumpulan indikator individu yang diperoleh dari survei untuk kasus diabetes. Dataset ini berguna untuk melakukan prediksi diabetes sehingga suatu individu dapat diketahui memiliki risiko tinggi diabetes atau tidak. Hal ini diperoleh dari fitur-fitur yang dapat dianalisis lebih lanjut sebelum mencapai kesimpulan.

Berikut adalah deskripsi singkat setiap kolom:

1. **HighBP**: Memiliki tekanan darah tinggi (BP: Blood Pressure) atau tidak
2. **HighChol**: Kolesterol tinggi atau tidak
3. **BMI**: Besaran Body Mass Index
4. **Smoker**: Perokok atau bukan perokok
5. **Stroke**: Pernah mengalami struk atau tidak
6. **HeartDiseaseorAttack**: Memiliki riwayat penyakit antara jantung koroner dan serangan jantung atau tidak sama sekali
7. **PhysActivity**: Aktif secara fisik dalam 30 hari terakhir atau tidak
8. **Fruits**: Mengonsumsi buah setiap hari atau tidak 
9. **Veggies**: Mengonsumsi sayur setiap hari atau tidak
10. **HvyAlcoholConsump**: Peminum berat alkohol atau bukan 
11. **AnyHealthcare**: Memiliki perlindungan kesehatan atau tidak, contohnya memiliki asuransi kesehatan
12. **GenHtlth**: Evaluasi mandiri terhadap kesehatan, skala 1-5 (1: Sangat baik, 2: Cukup Baik, 3: Baik, 4: Biasa saja, 5: Buruk)
13. **MentHlth**: Jumlah hari keadaan mental buruk dalam 30 hari terakhir (skala 0-30 hari)  
14. **PhysHlth**: Jumlah hari keadaan fisik buruk dalam 30 hari terakhir (skala 0-30 hari)
15. **DiffWalk**: Memiliki kesulitan berjalan atau menaiki tangga
16. **Sex**: (M) Male atau (F) Female
17. **Age**: 13 kategori umur (1: 18-24 tahun, 9: 60-64 tahun, 13: 80 tahun ke atas)
18. **Education**: Level edukasi skala 1-6 (1: Tidak pernah sekolah atau hanya TK, 2: SD, dst)
19. **Income**: Skala pendapatan 1-8
20. **Diabetes**: Apakah mengalami diabetes atau tidak (Kolom target)

In [92]:
# Import library di sini
import pandas as pd
from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.linear_model import LogisticRegression
from sklearn.preprocessing import LabelEncoder, MinMaxScaler
from sklearn.metrics import confusion_matrix, classification_report, f1_score, accuracy_score
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import StackingClassifier, RandomForestClassifier
from imblearn.under_sampling import RandomUnderSampler
from imblearn.over_sampling import RandomOverSampler

In [93]:
data = pd.read_csv("diabetes.csv")

In [94]:
X = data.drop(columns="Diabetes")
y = data["Diabetes"].copy()

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=123)
X_train, X_val, y_train, y_val = train_test_split(X_train, y_train, test_size=0.2, random_state=123)

# Gunakan data validasi untuk bereksperimen dengan model
# Gunakan data test untuk mengevaluasi model hanya di akhir eksperimen
df_train = pd.concat([X_train, y_train], axis=1)
df_val = pd.concat([X_val, y_val], axis=1)
df_test = pd.concat([X_test, y_test], axis=1)

_Deliverable_ yang akan dihasilkan adalah sebagai berikut:
1. berkas _notebook_ dengan format nama `PraktikumIF3270_M1_NIM1_NIM2.ipynb` untuk Bagian 1;
2. berkas _notebook_ dengan format nama `PraktikumIF3270_M2_NIM1_NIM2.ipynb` untuk Bagian 1 + Bagian 2; serta
3. berkas laporan dengan format nama `PraktikumIF3270_NIM1_NIM2.pdf` yang mencakup hal berikut:
    - hasil analisis data,
    - penanganan dari hasil analisis data,
    - justifikasi teknik-teknik yang dipilih,
    - perubahan yang dilakukan pada jawaban poin 1—5 jika ada,
    - desain eksperimen,
    - hasil eksperimen,
    - analisis dari hasil eksperimen,
    - kesimpulan,
    - pembagian tugas/kerja per anggota kelompok

Batas waktu pengumpulan:
- _Deliverable_ poin 1: Senin, 25 April 2023, pukul 12.00 WIB
- _Deliverable_ poin 2: Senin, 25 April 2023, pukul 21.00 WIB
- _Deliverable_ poin 3: Senin, 25 April 2023, pukul 21.00 WIB

## Bagian 1
---

### 1. Pembuatan Baseline Model
---
Buatlah _baseline_ dengan menggunakan model _logistic regression_.

In [95]:
# Check dataframe information
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50736 entries, 0 to 50735
Data columns (total 20 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   HighBP                50736 non-null  float64
 1   HighChol              50736 non-null  float64
 2   BMI                   50736 non-null  float64
 3   Smoker                50736 non-null  float64
 4   Stroke                50736 non-null  float64
 5   HeartDiseaseorAttack  50736 non-null  float64
 6   PhysActivity          50736 non-null  float64
 7   Fruits                50736 non-null  float64
 8   Veggies               50736 non-null  float64
 9   HvyAlcoholConsump     50736 non-null  float64
 10  AnyHealthcare         50736 non-null  float64
 11  GenHlth               50736 non-null  float64
 12  MentHlth              50736 non-null  float64
 13  PhysHlth              50736 non-null  float64
 14  DiffWalk              50736 non-null  float64
 15  Sex                

In [96]:
# Preprocessing non-numeric data
non_numeric_feature_names = data.select_dtypes(include=['object', 'bool']).columns.tolist()
print(f"Feature with non-numeric data type: {non_numeric_feature_names}")

label_encoder = LabelEncoder()
for feature in non_numeric_feature_names:
    df_train[feature] = label_encoder.fit_transform(df_train[feature])
    df_val[feature] = label_encoder.fit_transform(df_val[feature])
    df_test[feature] = label_encoder.fit_transform(df_test[feature])


X_train = df_train.drop(columns="Diabetes")
y_train = df_train["Diabetes"].copy()

X_val = df_val.drop(columns="Diabetes")
y_val = df_val["Diabetes"].copy()

X_test = df_test.drop(columns="Diabetes")
y_test = df_test["Diabetes"].copy()

Feature with non-numeric data type: ['Sex', 'Diabetes']


In [97]:
# Check preprocessing result
df_train

HighBP  HighChol   BMI  Smoker  Stroke  HeartDiseaseorAttack  \
21865     0.0       0.0  27.0     1.0     0.0                   0.0   
1925      0.0       0.0  27.0     0.0     0.0                   0.0   
49180     1.0       0.0  30.0     1.0     0.0                   0.0   
662       0.0       0.0  32.0     1.0     0.0                   0.0   
19989     1.0       1.0  24.0     1.0     0.0                   0.0   
...       ...       ...   ...     ...     ...                   ...   
46248     1.0       0.0  30.0     1.0     0.0                   0.0   
26462     0.0       0.0  28.0     0.0     0.0                   0.0   
28564     1.0       1.0  30.0     0.0     0.0                   0.0   
10746     1.0       0.0  37.0     1.0     0.0                   0.0   
21537     0.0       1.0  24.0     0.0     0.0                   0.0   

       PhysActivity  Fruits  Veggies  HvyAlcoholConsump  AnyHealthcare  \
21865           1.0     1.0      1.0                0.0            1.0   
1925            1.0     1.0      1.0                0.0            1.0   
49180           1.0     0.0      0.0                0.0            1.0   
662             1.0     0.0      1.0                0.0            1.0   
19989           1.0     1.0      1.0                0.0            1.0   
...             ...     ...      ...                ...            ...   
46248           1.0     1.0      1.0                0.0            1.0   
26462           1.0     1.0      0.0                0.0            0.0   
28564           1.0     1.0      1.0                0.0            1.0   
10746           0.0     0.0      1.0                0.0            0.0   
21537           1.0     1.0      1.0                0.0            1.0   

       GenHlth  MentHlth  PhysHlth  DiffWalk  Sex   Age  Education  Income  \
21865      1.0       0.0       0.0       1.0    1  10.0        5.0     8.0   
1925       2.0       3.0       0.0       0.0    1  12.0        6.0     8.0   
49180      3.0       7.0       3.0       1.0    0   6.0        6.0     7.0   
662        2.0       0.0       0.0       0.0    1   7.0        6.0     8.0   
19989      2.0       2.0      15.0       0.0    0  11.0        6.0     8.0   
...        ...       ...       ...       ...  ...   ...        ...     ...   
46248      3.0       0.0       0.0       0.0    1  11.0        5.0     8.0   
26462      1.0       0.0       0.0       0.0    1   6.0        6.0     8.0   
28564      2.0       0.0       0.0       0.0    1   6.0        6.0     7.0   
10746      4.0       5.0      30.0       1.0    0   9.0        4.0     6.0   
21537      1.0       2.0       0.0       0.0    1   3.0        6.0     8.0   

       Diabetes  
21865         0  
1925          0  
49180         0  
662           0  
19989         0  
...         ...  
46248         1  
26462         0  
28564         0  
10746         0  
21537         0  

[32470 rows x 20 columns]

In [98]:
# Logistic Regression Baseline Model Training
logreg = LogisticRegression(max_iter=1000, random_state=42)
logreg.fit(X_train, y_train)
y_pred = logreg.predict(X_val)

print("Classification Report:")
print(classification_report(y_val, y_pred))
print(f"F1 Score: {f1_score(y_val, y_pred)}")
print(f"Accuracy Score: {accuracy_score(y_val, y_pred)}")
print(f"Confusion Matrix:\n{confusion_matrix(y_val, y_pred)}")

Classification Report:
              precision    recall  f1-score   support

           0       0.88      0.98      0.93      7027
           1       0.56      0.15      0.24      1091

    accuracy                           0.87      8118
   macro avg       0.72      0.57      0.59      8118
weighted avg       0.84      0.87      0.84      8118

F1 Score: 0.24137931034482762
Accuracy Score: 0.8699186991869918
Confusion Matrix:
[[6894  133]
 [ 923  168]]


### 2. Analisis Data
---
Lakukan analisis data terkait hal berikut:
- _duplicate value_,
- _missing value_,
- _outlier_,
- _balance of data_.

#### _Duplicate Value_

In [99]:
duplicate = data[data.duplicated()]
print(f"Number of duplicate data rows: {len(duplicate)}")

Number of duplicate data rows: 2329


Dari eksekusi hasil kode bahwa terdapat cukup banyak baris pada dataset yang merupakan duplikat (memiliki nilai fitur yang sama) dari baris yang lain.

#### _Missing Value_

In [100]:
print("Missing value for each feature:")
print(data.isnull().sum().sort_values(ascending=False))

Missing value for each feature:
HighBP                  0
HighChol                0
Income                  0
Education               0
Age                     0
Sex                     0
DiffWalk                0
PhysHlth                0
MentHlth                0
GenHlth                 0
AnyHealthcare           0
HvyAlcoholConsump       0
Veggies                 0
Fruits                  0
PhysActivity            0
HeartDiseaseorAttack    0
Stroke                  0
Smoker                  0
BMI                     0
Diabetes                0
dtype: int64


Dari hasil eksekusi kode, tidak ditemukan data dengan _missing value_ pada dataset.

#### _Outlier_

In [101]:
numeric_feature_names = ['BMI', 'MentHlth', 'PhysHlth']

idx_to_remove = []
for feature in numeric_feature_names:
    Q1 = data[feature].quantile(0.25)
    Q3 = data[feature].quantile(0.75)
    IQR = Q3 - Q1

    lower_bound = Q1 - 1.5 * IQR
    upper_bound = Q3 + 1.5 * IQR

    outliers = data[(data[feature] < lower_bound) | (data[feature] > upper_bound)][feature]

    print(f"Feature '{feature}': {len(outliers)} outliers")
    if len(outliers) != 0:
        print(f"Values: {set(outliers.tolist())}")
        idx_to_remove.extend(outliers.index.tolist())
    print()



Feature 'BMI': 1979 outliers
Values: {13.0, 42.0, 43.0, 44.0, 45.0, 46.0, 47.0, 48.0, 49.0, 50.0, 51.0, 52.0, 53.0, 54.0, 55.0, 56.0, 57.0, 58.0, 59.0, 60.0, 61.0, 62.0, 63.0, 64.0, 65.0, 66.0, 67.0, 68.0, 70.0, 71.0, 72.0, 73.0, 74.0, 75.0, 76.0, 77.0, 79.0, 80.0, 81.0, 82.0, 83.0, 84.0, 87.0, 89.0, 92.0, 95.0, 98.0}

Feature 'MentHlth': 7308 outliers
Values: {6.0, 7.0, 8.0, 9.0, 10.0, 11.0, 12.0, 13.0, 14.0, 15.0, 16.0, 17.0, 18.0, 19.0, 20.0, 21.0, 22.0, 23.0, 24.0, 25.0, 26.0, 27.0, 28.0, 29.0, 30.0}

Feature 'PhysHlth': 8198 outliers
Values: {8.0, 9.0, 10.0, 11.0, 12.0, 13.0, 14.0, 15.0, 16.0, 17.0, 18.0, 19.0, 20.0, 21.0, 22.0, 23.0, 24.0, 25.0, 26.0, 27.0, 28.0, 29.0, 30.0}



Dari hasil eksekusi kode, ditemukan cukup banyak _outlier_ pada fitur-fitur dengan tipe data numerik.

#### _Balance of Data_

In [102]:
diabetes_true = data[data["Diabetes"] == True]
diabetes_false = data[data["Diabetes"] == False]
print(f"Number of rows with diabetes = true: {len(diabetes_true)}")
print(f"Number of rows with diabetes = false: {len(diabetes_false)}")
print(f"Proportion of minority over dataset: {len(diabetes_true)/len(data)}")

Number of rows with diabetes = true: 6946
Number of rows with diabetes = false: 43790
Proportion of minority over dataset: 0.13690476190476192


Dilihat dari hasil eksekusi kode, maka dapat dikatakan bahwa dataset bersifat _imbalanced_ karena jumlah baris dengan nilai kolom target yaitu Diabetes = true hanya sekitar 13.69% dari keseluruhan data.

### 3. Rencana Penanganan
---
Jelaskan rencana penanganan yang ada pada poin 2.

#### _Duplicate Value_

Terdapat beberapa alasan mengapa data duplikat harus ditangani:
1. Data duplikat dapat memengaruhi proses pembelajaran atau _training_ dari suatu model karena menambahkan pola yang sama dengan data duplikat lainnya. Hal ini dapat membuat model yang dilatih lebih fokus terhadap pola-pola yang tidak representatif terhadap dataset sebenarnya.
2. Data duplikat dapat memengaruhi proses evaluasi model, model tentunya memberikan prediksi yang sama untuk data-data duplikat. Akibatnya, hasil prediksi data duplikat memberikan bobot yang lebih besar terhadap perhitungan nilai metrik, meskipun data tersebut tidak representatif.
3. Data duplikat meningkatkan beban komputasi saat melatih model, karena model memproses data yang sama berulang kali.

Oleh karena itu, untuk menangani data duplikat, data tersebut akan dihapus dari dataset.

#### _Missing Value_

Terdapat beberapa alasan mengapa data dengan _missing value_ harus ditangani:
1. Data dengan _missing value_ memengaruhi nilai analisis statistik pada data seperti mean dan median dengan memberikan bias dan menyebabkan nilai tidak akurat.
2. Data dengan _missing value_, jika dibiarkan akan menyebabkan distorsi hingga kegagalan dalam proses pembelajaran model. Hal ini karena model tidak dapat secara langsung menangani _missing value_.
3. Data dengan _missing value_, jika ditangani dengan baik akan mampu meningkatkan efektivitas dari model yang dilatih.

Jika terdapat _missing value_, maka rencana penanganannya adalah menghapus data dengan _missing value_ dari dataset. Rencana ini dipilih daripada mengisi _missing value_ dengan nilai tertentu. Hal ini demi kemudahan dan efisiensi. Selain itu, mengisi _missing value_ dengan nilai tertentu juga ditakutkan mengubah pola atau representasi dari dataset.
Berdasarkan hasil analisis data, dataset tidak memiliki _missing value_, sehingga penanganan tidak perlu dilakukan.

#### _Outlier_
Terdapat beberapa alasan mengapa data _outlier_ harus ditangani:
1. Data _outlier_ dapat menyebabkan terjadinya bias terhadap nilai analisis statistik seperti mean dan standar deviasi yang menyebabkan dataset tidak tercerminkan dengan benar.
2. Data _outlier_ dapat memengaruhi proses pembelajaran model karena memberikan bobot kepada data yang sebenarnya tidak representatif. Jika _outlier_ cukup banyak, dapat menyebabkan model tidak stabil dan prediksi tidak akurat.

Oleh karena itu, untuk menangani _outlier_, data dengan _outlier_ akan dihapus dari dataset.

#### _Balance of Data_

Terdapat beberapa alasan mengapa data yang tidak seimbang harus ditangani:
1. _Imbalanced dataset_ dapat menyebabkan model yang dilatih cenderung untuk meminimalkan kesalahan prediksi kelas mayoritas dibandingkan dengan kelas minoritas. Hal ini menyebabkan model kurang baik atau bahkan buruk dalam memprediksi kelas minoritas.
2. Metrik evaluasi yang digunakan bisa jadi tidak bekerja dengan baik atau tidak memberikan informasi yang tepat ketika kelas minoritas menjadi kelas yang penting untuk dapat diprediksi dengan benar.

Untuk menangani _imbalanced dataset_, penanganan yang akan dilakukan adalah melakukan undersampling terhadap data kelas mayoritas memanfaatkan RandomUnderSampler untuk menyeimbangkan jumlah data kedua kelas. Teknik ini dirasa lebih baik pada dataset ini dibandingkan _oversampling_ yang menciptakan data minoritas sintetis yang jumlahnya cukup banyak agar seimbang dengan kelas mayoritas


### 4. Teknik _Encoding_
---
Jelaskan teknik _encoding_ yang digunakan terhadap data yang disediakan apabila dilakukan, disertai dengan alasan.

Pada dataset ini, terdapat 1 fitur dengan tipe data kategorikal (object), yaitu 'Sex', yang memiliki dua nilai: F (Female) dan M (Male), dan 1 fitur dengan tipe data kategorikal (boolean), yaitu 'Diabetes', yang memiliki dua nilai: True dan False. Untuk melakukan pemrosesan data dalam model nantinya, kedua fitur ini perlu dikonversi ke format numerik. Pendekatan yang dilakukan adalah Label Encoding, yang cocok untuk data dengan sifat biner. Label Encoding memungkinkan fitur-fitur tersebut dikonversi menjadi angka, di mana setiap kategori akan diberikan label numerik tertentu.

### 5. Desain Eksperimen 
---
Buatlah desain eksperimen dengan menentukan hal berikut: 
- tujuan eksperimen,
- variabel dependen dan independen,
- strategi eksperimen,
- skema validasi.

#### Tujuan Eksperimen
Eksperimen ini bertujuan untuk mencoba berbagai parameter, metode pemrosesan, dan berbagai jenis algoritma pembelajaran untuk menghasilkan model yang paling efektif dan akurat dalam melakukan prediksi terhadap dataset Diabetes Health Indicator. Keefektifan dan keakuratan dari model akan dievaluasi dengan memanfaatkan metrik berupa nilai akurasi, f1, _confusion matrix_, serta _classification report_.


#### Variabel Dependen dan Independen
Variabel dependen pada eksperimen ini adalah variabel yang akan diprediksi, yaitu variabel Diabetes yang akan bernilai True atau False. Sedangkan, variabel independen adalah variabel yang akan dijadikan input untuk prediksi model. Variabel independen akan memengaruhi variabel dependen dan digunakan untuk memprediksi nilai variabel dependen. Dalam eksperimen ini, variabel independen adalah HighBP, HighChol, BMI, Smoker, Stroke, HeartDiseaseorAttack, PhysActivity, Fruits, Veggies, HvyAlcoholConsump, AnyHealthcare, GenHlth, MentHlth, PhysHlth, DiffWalk, Sex, Age, Education, Income.

#### Strategi Eksperimen

Berikut merupakan langkah-langkah eksperimen yang akan dilakukan:

##### 1. Melakukan Penanganan Data
Sebelum memulai eksperimen, dataset akan ditangani terlebih dahulu dengan melakukan rencana penanganan yang telah ditulis pada bagian sebelumnya. Rencana penanganan untuk _imbalanced dataset_ akan dilakukan setelah membuang beberapa fitur data latih. 
Oleh karena itu akan dilakukan:
- Penghapusan data duplikat
- Penghapusan data _outlier_

Setelah penghapusan data, data numerik juga akan diolah dan di-_scaling_ dengan menggunakan MinMaxScaler. Hal ini penting dilakukan untuk mencegah bobot antar fitur yang tidak seimbang serta konvergensi yang lebih cepat pada beberapa model klasifikasi seperti Logistic Regression. Teknik _encoding_ yang telah dijelaskan juga akan dilakukan pada beberapa data kategorikal.

##### 2. Membuang Beberapa Fitur Data Latih
Sebagai bagian dari eksperimen, akan dilakukan pemilihan terhadap fitur data latih yang dimasukkan ke dalam model. Hal ini dilakukan dengan membuang beberapa fitur yang dianggap tidak diperlukan atau tidak berhubungan dengan target prediksi.

##### 3. Melakukan Penanganan _Imbalanced Data_
Sebagai bagian dari eksperimen, dataset kemudian akan ditangani ketidakseimbangannya dengan melakukan _undersampling_ memanfaatkan RandomUnderSampler. Teknik _undersampling_ dipilih untuk mengurangi jumlah kelas mayoritas sehingga seimbang dengan kelas minoritas.

##### 4. Melakukan Eksperimen Terhadap Berbagai Jenis Model
Selanjutnya, akan dilakukan eksperimen pelatihan berbagai jenis model dan konfigurasi. Eksperimen yang akan dilakukan adalah
- Penggantian model klasifikasi: Beberapa model klasifikasi seperti Logistic Regression, Random Forest, KNN, dan Decision Tree akan dicoba untuk pelatihan. Hyperparameter tuning dengan grid search akan dilakukan terhadap berbagai model tersebut.
- Beberapa model klasifikasi akan di _stack_ untuk menghasilkan model baru.

##### 5. Menghasilkan Metrik Evaluasi Model
Setiap model yang dihasilkan kemudian akan divalidasi dan dievaluasi dengan mengeluarkan metrik berupa _classification report_, _confusion matrix_, _f1 score_, dan _accuracy score_.


#### Skema Validasi
Skema validasi yang kami gunakan adalah hold-out validation. Skema ini melibatkan pemisahan dataset menjadi dua subset terpisah: satu untuk pelatihan (training set) dan satu lagi untuk pengujian (test set). Skema ini dipilih karena memungkinkan evaluasi yang objektif dan tidak bias terhadap model yang dilatih dengan training set.

Kami menggunakan metrik F1-score sebagai indikator utama performa model karena F1-score menggunakan presisi dan recall. Metrik ini sangat berguna dalam konteks data yang tidak seimbang, karena dapat menyeimbangkan kepentingan antara presisi dan kemampuan model untuk mendeteksi semua instance positif (recall).


## Bagian 2
---

### 6. Proses Eksperimen
---
Implementasikan strategi eksperimen dan skema validasi yang telah ditentukan pada poin 5.

#### Penanganan data
Menghapus baris, melakukan _scaling_, serta menerapkan teknik _encoding_.

In [103]:
# Uncommon value of each categorical feature
threshold = 0.005

for feature in data.columns.to_list():
    if feature not in numeric_feature_names:
        value_counts = data[feature].value_counts(normalize=True)
        uncommon_values = value_counts[value_counts < threshold].index.tolist()

        uncommon_values_column = data[data[feature].isin(uncommon_values)][feature]

        print(f"Feature '{feature}': {len(uncommon_values)} uncommmon values")
        if len(uncommon_values) != 0:
            print(f"Values: {uncommon_values}")
            idx_to_remove.extend(uncommon_values_column.index.tolist())
        print()

Feature 'HighBP': 0 uncommmon values

Feature 'HighChol': 0 uncommmon values

Feature 'Smoker': 0 uncommmon values

Feature 'Stroke': 0 uncommmon values

Feature 'HeartDiseaseorAttack': 0 uncommmon values

Feature 'PhysActivity': 0 uncommmon values

Feature 'Fruits': 0 uncommmon values

Feature 'Veggies': 0 uncommmon values

Feature 'HvyAlcoholConsump': 0 uncommmon values

Feature 'AnyHealthcare': 0 uncommmon values

Feature 'GenHlth': 0 uncommmon values

Feature 'DiffWalk': 0 uncommmon values

Feature 'Sex': 0 uncommmon values

Feature 'Age': 0 uncommmon values

Feature 'Education': 1 uncommmon values
Values: [1.0]

Feature 'Income': 0 uncommmon values

Feature 'Diabetes': 0 uncommmon values



In [104]:
# Remove row with outlier and uncommon values
print(f"Shape before removing outliers and uncommon values: {data.shape}")
data = data.drop(idx_to_remove)
print(f"Shape after removing outliers and uncommon values: {data.shape}")

Shape before removing outliers and uncommon values: (50736, 20)
Shape after removing outliers and uncommon values: (37301, 20)


In [105]:
# Remove duplicate rows
print(f"Shape before removing duplicate rows: {data.shape}")
data = data.drop_duplicates()
print(f"Shape after removing duplicate rows: {data.shape}")

Shape before removing duplicate rows: (37301, 20)
Shape after removing duplicate rows: (34974, 20)


In [106]:
# Scale non-categorical features
scaler = MinMaxScaler()
numeric_feature_names.extend(['GenHlth', 'Age', 'Education', 'Income'])
for feature in numeric_feature_names:
    data[feature] = scaler.fit_transform(data[[feature]])

In [107]:
# Encode categorical features
label_encoder = LabelEncoder()

for feature in non_numeric_feature_names:
    data[feature] = label_encoder.fit_transform(data[feature])

In [108]:
# Show dataframe
data[data['GenHlth'] == 5.0]

Empty DataFrame
Columns: [HighBP, HighChol, BMI, Smoker, Stroke, HeartDiseaseorAttack, PhysActivity, Fruits, Veggies, HvyAlcoholConsump, AnyHealthcare, GenHlth, MentHlth, PhysHlth, DiffWalk, Sex, Age, Education, Income, Diabetes]
Index: []

#### Pemilihan Fitur
Membuang beberapa fitur yang dianggap tidak diperlukan atau tidak berhubungan dengan target prediksi

Setelah beberapa percobaan dilakukan, diputuskan untuk tidak membuang fitur apapun dari dataset karena memberikan metrik yang paling baik.

#### Penanganan _Imbalanced Data_
Melakukan teknik _undersampling_ untuk menyeimbangkan data.

In [109]:
# Initial train-validation-test split
X = data.drop(columns="Diabetes")
y = data["Diabetes"].copy()

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)
X_train, X_val, y_train, y_val = train_test_split(X_train, y_train, test_size=0.2, random_state=42)


In [110]:
# Handling with RandomUnderSampler
undersampler = RandomUnderSampler(sampling_strategy="majority", random_state=42)
undersampler_X_train, undersampler_y_train = undersampler.fit_resample(X_train, y_train)

#### Eksperimen Model
Mencoba berbagai model klasifikasi, melakukan _hyperparameter tuning_, serta melakukan _stacking_ model.


Model _Logistic Regression_ dengan _Hyperparameter Tuning_

In [111]:
logreg = LogisticRegression()
param_grid = {
    'C': [0.001, 0.01, 0.1, 1, 10, 100],
    'penalty': ['l1', 'l2'],
    'solver': ['liblinear', 'saga']
}
grid_search = GridSearchCV(estimator=logreg, param_grid=param_grid, cv=5)
    
grid_search.fit(undersampler_X_train, undersampler_y_train)
logreg_best_param = grid_search.best_params_
logreg_best_score = grid_search.best_score_

best_logreg = grid_search.best_estimator_
logreg_y_pred = best_logreg.predict(X_val)

print("Classification Report:")
print(classification_report(y_val, logreg_y_pred))
print(f"F1 Score: {f1_score(y_val, logreg_y_pred)}")
print(f"Accuracy Score: {accuracy_score(y_val, logreg_y_pred)}")
print(f"Confusion Matrix:\n{confusion_matrix(y_val, logreg_y_pred)}")
print()
print(f"Grid Search Best Parameters: {logreg_best_param}")
print(f"Grid Search Best Score: {logreg_best_score}")

Classification Report:
              precision    recall  f1-score   support

           0       0.95      0.73      0.83      4934
           1       0.27      0.74      0.39       662

    accuracy                           0.73      5596
   macro avg       0.61      0.73      0.61      5596
weighted avg       0.87      0.73      0.78      5596

F1 Score: 0.3929288871032544
Accuracy Score: 0.7299857040743388
Confusion Matrix:
[[3596 1338]
 [ 173  489]]

Grid Search Best Parameters: {'C': 10, 'penalty': 'l2', 'solver': 'liblinear'}
Grid Search Best Score: 0.738979508391575


Model _Random Forest_ dengan _Hyperparameter Tuning_

In [112]:
random_forest = RandomForestClassifier()
param_grid = {
    'n_estimators': [30, 50, 100], 
    'min_samples_split': [2, 5, 10],
    'min_samples_leaf': [1, 2, 4],  
}
grid_search = GridSearchCV(estimator=random_forest, param_grid=param_grid, cv=5)
    
grid_search.fit(undersampler_X_train, undersampler_y_train)
rf_best_param = grid_search.best_params_
rf_best_score = grid_search.best_score_

best_random_forest = grid_search.best_estimator_
rf_y_pred = best_random_forest.predict(X_val)

print("Classification Report:")
print(classification_report(y_val, rf_y_pred))
print(f"F1 Score: {f1_score(y_val, rf_y_pred)}")
print(f"Accuracy Score: {accuracy_score(y_val, rf_y_pred)}")
print(f"Confusion Matrix:\n{confusion_matrix(y_val, rf_y_pred)}")
print()
print(f"Grid Search Best Parameters: {rf_best_param}")
print(f"Grid Search Best Score: {rf_best_score}")

Classification Report:
              precision    recall  f1-score   support

           0       0.95      0.70      0.80      4934
           1       0.25      0.75      0.37       662

    accuracy                           0.70      5596
   macro avg       0.60      0.72      0.59      5596
weighted avg       0.87      0.70      0.75      5596

F1 Score: 0.3724604966139955
Accuracy Score: 0.7019299499642602
Confusion Matrix:
[[3433 1501]
 [ 167  495]]

Grid Search Best Parameters: {'min_samples_leaf': 4, 'min_samples_split': 10, 'n_estimators': 50}
Grid Search Best Score: 0.7380027575714353


Model Decision Tree

In [113]:
dt = DecisionTreeClassifier()
param_grid = {
    'max_depth': [None, 10, 20, 30],
    'min_samples_split': [2, 5, 10], 
    'min_samples_leaf': [1, 2, 4], 
    'criterion': ['gini', 'entropy'] 
}
grid_search = GridSearchCV(estimator=dt, param_grid=param_grid, cv=5)
    
grid_search.fit(undersampler_X_train, undersampler_y_train)
dt_best_param = grid_search.best_params_
dt_best_score = grid_search.best_score_

best_dt = grid_search.best_estimator_
dt_y_pred = best_random_forest.predict(X_val)

print("Classification Report:")
print(classification_report(y_val, dt_y_pred))
print(f"F1 Score: {f1_score(y_val, dt_y_pred)}")
print(f"Accuracy Score: {accuracy_score(y_val, dt_y_pred)}")
print(f"Confusion Matrix:\n{confusion_matrix(y_val, dt_y_pred)}")
print()
print(f"Grid Search Best Parameters: {dt_best_param}")
print(f"Grid Search Best Score: {dt_best_score}")

Classification Report:
              precision    recall  f1-score   support

           0       0.95      0.70      0.80      4934
           1       0.25      0.75      0.37       662

    accuracy                           0.70      5596
   macro avg       0.60      0.72      0.59      5596
weighted avg       0.87      0.70      0.75      5596

F1 Score: 0.3724604966139955
Accuracy Score: 0.7019299499642602
Confusion Matrix:
[[3433 1501]
 [ 167  495]]

Grid Search Best Parameters: {'criterion': 'gini', 'max_depth': 10, 'min_samples_leaf': 4, 'min_samples_split': 10}
Grid Search Best Score: 0.6931370703180715


Model _Stacking Ensemble_ dengan _Logistic Regression_ dan _Random Forest_

In [114]:
lr = LogisticRegression(solver="liblinear", random_state=42)
rf = RandomForestClassifier(criterion='entropy', random_state=42)
final_estimator = LogisticRegression(solver="liblinear", max_iter=1000, random_state=42)

stacking_ensemble = StackingClassifier(
    estimators=[
        ("lr_1", lr),
        ("rf_1", rf)
    ],
    final_estimator=final_estimator
)
stacking_ensemble.fit(undersampler_X_train, undersampler_y_train)
stacking_ensemble_pred = stacking_ensemble.predict(X_val)

print("Classification Report:")
print(classification_report(y_val, stacking_ensemble_pred))
print(f"F1 Score: {f1_score(y_val, stacking_ensemble_pred)}")
print(f"Accuracy Score: {accuracy_score(y_val, stacking_ensemble_pred)}")
print(f"Confusion Matrix:\n{confusion_matrix(y_val, stacking_ensemble_pred)}")

Classification Report:
              precision    recall  f1-score   support

           0       0.95      0.73      0.83      4934
           1       0.27      0.73      0.39       662

    accuracy                           0.73      5596
   macro avg       0.61      0.73      0.61      5596
weighted avg       0.87      0.73      0.78      5596

F1 Score: 0.39176090468497576
Accuracy Score: 0.7308791994281629
Confusion Matrix:
[[3605 1329]
 [ 177  485]]


### 7. Analisis & Kesimpulan
---
Berdasarkan hasil prediksi yang dihasilkan, buatlah kesimpulan analisis **hasil diabetes**.

Dilihat dari hasil eksperimen yang telah dilakukan diatas, dapat disimpulkan bahwa model dengan tingkat akurasi terbaik (dilihat dari F1 Score) adalah Model _Logistic Regression_ dengan _Hyperparameter Tuning_. Hasil akhir menunjukkan bahwa Model Logistic Regression yang telah dioptimasi dengan Hyperparameter Tuning mencapai skor F1 sebesar 0.3929, yang menandakan kemampuan model untuk mencapai keseimbangan antara _precision_ dan _recall_.

In [115]:
test_y_pred = best_logreg.predict(X_test)

print("Classification Report:")
print(classification_report(y_test, test_y_pred))
print(f"F1 Score: {f1_score(y_test, test_y_pred)}")
print(f"Accuracy Score: {accuracy_score(y_test, test_y_pred)}")
print(f"Confusion Matrix:\n{confusion_matrix(y_test, test_y_pred)}")

Classification Report:
              precision    recall  f1-score   support

           0       0.96      0.72      0.82      6214
           1       0.25      0.75      0.38       781

    accuracy                           0.72      6995
   macro avg       0.60      0.73      0.60      6995
weighted avg       0.88      0.72      0.77      6995

F1 Score: 0.37645536869340235
Accuracy Score: 0.7243745532523231
Confusion Matrix:
[[4485 1729]
 [ 199  582]]


In [119]:
for i in range(len(y_test)):
    print(f"Actual: {y_test.iloc[i] if hasattr(y_test, 'iloc') else y_test[i]} - Predicted: {test_y_pred[i]}")


Actual: 0 - Predicted: 1
Actual: 0 - Predicted: 1
Actual: 1 - Predicted: 1
Actual: 0 - Predicted: 1
Actual: 0 - Predicted: 1
Actual: 0 - Predicted: 0
Actual: 0 - Predicted: 1
Actual: 0 - Predicted: 0
Actual: 0 - Predicted: 1
Actual: 0 - Predicted: 1
Actual: 1 - Predicted: 0
Actual: 0 - Predicted: 1
Actual: 1 - Predicted: 0
Actual: 0 - Predicted: 0
Actual: 0 - Predicted: 0
Actual: 0 - Predicted: 0
Actual: 0 - Predicted: 0
Actual: 0 - Predicted: 0
Actual: 0 - Predicted: 0
Actual: 0 - Predicted: 0
Actual: 0 - Predicted: 0
Actual: 0 - Predicted: 0
Actual: 0 - Predicted: 0
Actual: 0 - Predicted: 0
Actual: 0 - Predicted: 0
Actual: 0 - Predicted: 0
Actual: 0 - Predicted: 0
Actual: 0 - Predicted: 0
Actual: 0 - Predicted: 1
Actual: 0 - Predicted: 0
Actual: 0 - Predicted: 0
Actual: 0 - Predicted: 0
Actual: 0 - Predicted: 0
Actual: 0 - Predicted: 1
Actual: 0 - Predicted: 0
Actual: 0 - Predicted: 0
Actual: 0 - Predicted: 0
Actual: 0 - Predicted: 0
Actual: 0 - Predicted: 0
Actual: 0 - Predicted: 1


Berdasarkan analisis hasil prediksi, model sudah mampu mengklasifikasikan sebagian besar baris data secara akurat, namun masih terdapat beberapa data yang belum dapat diklasifikasikan dengan tepat. Kondisi ini dapat diakibatkan oleh beberapa faktor, antara lain: model klasifikasi yang belum sepenuhnya sesuai, metode pre-processing data yang kurang optimal, atau strategi penanganan imbalance dataset yang belum tepat. Meskipun begitu, dengan mempertimbangkan hasil eksperimen yang kami lakukan, model klasifikasi _Logistic Regression_ dengan _Hyperparameter Tuning_ memberikan hasil dengan metrik yang lebih baik dibandingkan model klasifikasi lainnya yang kami uji, seperti _Random Forest_ dengan _Hyperparameter Tuning_, _Decision Tree_, serta _Stacking Ensemble_ dengan _Logistic Regression_ dan _Random Forest_.